### ISSUE TITLE
inconsistencies in OpenSSL-1.1.0 builds using Mingw-w64 and MSVC

### ISSUE THREAD

**Issue by sav-ix** [openssl/openssl#1105]
Compared recent OpenSSL builds using Mings-64 and MSVC and got feedback on PRs: https://github.com/openssl/openssl/commit/e990ec5234d9daad66359833c40e4536d7fce499 , https://github.com/openssl/openssl/commit/2bc57c88d852d83ee31f68f5b84b36d54b4c3dc4 and not only.

- Mingw-w64 installation contain files:
```
bin/c_rehash
ssl/misc/c_hash
ssl/misc/c_info
ssl/misc/c_issuer
ssl/misc/c_name
ssl/misc/CA.pl
ssl/misc/tsget
```
which are missing in MSVC installation,

- Html docs in Mingw-w64 installation contain files:
```
share\doc\openssl\html\man1\c_rehash.html
share\doc\openssl\html\man3\CRYPTO_malloc.html
share\doc\openssl\html\man3\OPENSSL_cleanse.html
```
which are missing in MSVC html docs,

- Html docs in MSVC installation contain files:
```
html\man1\contributedundertheOpenSSLlicense.c_rehash.html
html\man1\OriginaltextbyJamesWestby.html
html\man3\OCSP_RESPONSE_free.html
html\man3\OCSP_response_create.html
html\man3\OCSP_request_sign.html
html\man3\OCSP_request_onereq_get0.html
html\man3\OCSP_REQUEST_free.html
html\man3\OCSP_request_onereq_count.html
html\man3\OCSP_request_add1_cert.html
html\man3\OCSP_request_add0_id.html
html\man3\OCSP_id_issuer_cmp.html
html\man3\OCSP_id_get0_info.html
html\man3\OCSP_id_cmp.html
html\man3\OCSP_copy_nonce.html
html\man3\OCSP_check_validity.html
html\man3\OCSP_check_nonce.html
html\man3\OCSP_cert_id_new.html
html\man3\OCSP_CERTID_free.html
html\man3\OPENSSL_cleanseCRYPTO_malloc.html
html\man3\OCSP_resp_get0.html
html\man3\OCSP_response_get1_basic.html
html\man3\OCSP_single_get0_status.html
html\man3\OCSP_resp_find.html
html\man3\OCSP_resp_count.html
html\man3\OCSP_resp_find_status.html
html\man3\OCSP_basic_add1_nonce.html
```
which are missing in Mingw-w64 html docs,

- All files in html docs directory of Mingw-w64 installation contain empty tags:
```
<title></title>
```

while all files in html docs of MSVC installation contain appropriate text inside that tags,

- Some files in html docs of Mingw-w64 installation contain plain text, eg:
```
<pre><code> #include &lt;openssl/evp.h&gt;

 EVP_CIPHER_CTX *EVP_CIPHER_CTX_new(void);
```

while the same text in html docs of MSVC installation would be:
```
<pre><code> <span class="comment">#include &lt;openssl/evp.h&gt;</span>
 
 <span class="variable">EVP_CIPHER_CTX</span> <span class="variable">*EVP_CIPHER_CTX_new</span><span class="operator">(</span><span class="variable">void</span><span class="operator">);</span>
```

- MSVC builds doesn't utilize all CPU cores during documentation build, unlike during binaries build using MSVC, or documentation build using Mingw-w64,

- Noise in stderr still exists:
```
 	link /nologo /debug /dll  /implib:libcrypto.lib /out:libcrypto-1_1-x64.dll /def:libcrypto-1_1-x64.def @C:\Users\user\AppData\Local\Temp\nm8719.tmp || (DEL /Q libcrypto.* libcrypto-1_1-x64.* && EXIT 1)
   Creating library libcrypto.lib and object libcrypto.exp
	DEL /Q /F apps\libcrypto-1_1-x64.dll
Could Not Find c:\libOPENSSL-1.1.0-dev\src\apps\libcrypto-1_1-x64.dll
	DEL /Q /F test\libcrypto-1_1-x64.dll
Could Not Find c:\libOPENSSL-1.1.0-dev\src\test\libcrypto-1_1-x64.dll
	COPY libcrypto-1_1-x64.dll apps
        1 file(s) copied.

[snip]

   Creating library libssl.lib and object libssl.exp
	DEL /Q /F apps\libssl-1_1-x64.dll
Could Not Find c:\libOPENSSL-1.1.0-dev\src\apps\libssl-1_1-x64.dll
	DEL /Q /F test\libssl-1_1-x64.dll
Could Not Find c:\libOPENSSL-1.1.0-dev\src\test\libssl-1_1-x64.dll
	COPY libssl-1_1-x64.dll apps
        1 file(s) copied.
```

Dunno whether I was wrong, when reported in https://github.com/openssl/openssl/issues/1072, that it was fixed, or it re-appeared because of some reason. Since `/Q` switch is not enough to silence errors, the possible solution is to replace `DEL ...dll` commands with condition `if exist ...dll (DEL ...dll)`.

Some of issues above could describe expected behavior. In such case no need to reply on them.

**Comment by levitte** [comment]
> Compared recent OpenSSL builds using Mings-64 and MSVC and got feedback on PRs: https://github.com/openssl/openssl/commit/e990ec5234d9daad66359833c40e4536d7fce499 , https://github.com/openssl/openssl/commit/2bc57c88d852d83ee31f68f5b84b36d54b4c3dc4 and not only.

> - Mingw-w64 installation contain files:

>   ```
>   bin/c_rehash
>   ssl/misc/c_hash
>   ssl/misc/c_info
>   ssl/misc/c_issuer
>   ssl/misc/c_name
>   ssl/misc/CA.pl
>   ssl/misc/tsget
>   ```
>   which are missing in MSVC installation,

Yup.  Most of those are shell scripts, so completely unusable on native Windows.  CA.pl is the exception, but is there a way to associate it with perl so it becomes easy to use on the command line?

> - Html docs in Mingw-w64 installation contain files:

>   ```
>   share\doc\openssl\html\man1\c_rehash.html
>   share\doc\openssl\html\man3\CRYPTO_malloc.html
>   share\doc\openssl\html\man3\OPENSSL_cleanse.html
>   ```
>   which are missing in MSVC html docs,
>   
> - Html docs in MSVC installation contain files:

>   ```
>   html\man1\contributedundertheOpenSSLlicense.c_rehash.html
>   html\man1\OriginaltextbyJamesWestby.html
>   html\man3\OCSP_RESPONSE_free.html
>   html\man3\OCSP_response_create.html
>   html\man3\OCSP_request_sign.html
>   html\man3\OCSP_request_onereq_get0.html
>   html\man3\OCSP_REQUEST_free.html
>   html\man3\OCSP_request_onereq_count.html
>   html\man3\OCSP_request_add1_cert.html
>   html\man3\OCSP_request_add0_id.html
>   html\man3\OCSP_id_issuer_cmp.html
>   html\man3\OCSP_id_get0_info.html
>   html\man3\OCSP_id_cmp.html
>   html\man3\OCSP_copy_nonce.html
>   html\man3\OCSP_check_validity.html
>   html\man3\OCSP_check_nonce.html
>   html\man3\OCSP_cert_id_new.html
>   html\man3\OCSP_CERTID_free.html
>   html\man3\OPENSSL_cleanseCRYPTO_malloc.html
>   html\man3\OCSP_resp_get0.html
>   html\man3\OCSP_response_get1_basic.html
>   html\man3\OCSP_single_get0_status.html
>   html\man3\OCSP_resp_find.html
>   html\man3\OCSP_resp_count.html
>   html\man3\OCSP_resp_find_status.html
>   html\man3\OCSP_basic_add1_nonce.html
>   ```
>   which are missing in Mingw-w64 html docs,

Interesting...  I'll check that more closely.

> - All files in html docs directory of Mingw-w64 installation contain empty tags:

>   ```
>   <title></title>
>   ```

>   while all files in html docs of MSVC installation contain appropriate text inside that tags,

Ah.  Will fix!

> - Some files in html docs of Mingw-w64 installation contain plain text, eg:

>   ```
>   <pre><code> #include &lt;openssl/evp.h&gt;
>   
>    EVP_CIPHER_CTX *EVP_CIPHER_CTX_new(void);
>   ```

>   while the same text in html docs of MSVC installation would be:

>   ```
>   <pre><code> <span class="comment">#include &lt;openssl/evp.h&gt;</span>
>    
>    <span class="variable">EVP_CIPHER_CTX</span> <span class="variable">*EVP_CIPHER_CTX_new</span><span class="operator">(</span><span class="variable">void</span><span class="operator">);</span>
>   ```

That's a difference that we don't control.  I believe it comes down to different perl versions, or that some perl dists add the extra `<span>` tags (I use perl 5.22 on Linux and haven't seen those).

> - MSVC builds doesn't utilize all CPU cores during documentation build, unlike during binaries build using MSVC, or documentation build using Mingw-w64,

Will not fix at this stage, that would require a too big change in the configuration and makefile builder.  I plan on making such a change for a future OpenSSL, probably 1.2.0.

> - Noise in stderr still exists:

>   ```
>    	link /nologo /debug /dll  /implib:libcrypto.lib /out:libcrypto-1_1-x64.dll /def:libcrypto-1_1-x64.def @C:\Users\user\AppData\Local\Temp\nm8719.tmp || (DEL /Q libcrypto.* libcrypto-1_1-x64.* && EXIT 1)
>      Creating library libcrypto.lib and object libcrypto.exp
>   	DEL /Q /F apps\libcrypto-1_1-x64.dll
>   Could Not Find c:\libOPENSSL-1.1.0-dev\src\apps\libcrypto-1_1-x64.dll
>   	DEL /Q /F test\libcrypto-1_1-x64.dll
>   Could Not Find c:\libOPENSSL-1.1.0-dev\src\test\libcrypto-1_1-x64.dll
>   	COPY libcrypto-1_1-x64.dll apps
>           1 file(s) copied.
>   
>   [snip]
>   
>      Creating library libssl.lib and object libssl.exp
>   	DEL /Q /F apps\libssl-1_1-x64.dll
>   Could Not Find c:\libOPENSSL-1.1.0-dev\src\apps\libssl-1_1-x64.dll
>   	DEL /Q /F test\libssl-1_1-x64.dll
>   Could Not Find c:\libOPENSSL-1.1.0-dev\src\test\libssl-1_1-x64.dll
>   	COPY libssl-1_1-x64.dll apps
>           1 file(s) copied.
>   ```

>   Dunno whether I was wrong, when reported in https://github.com/openssl/openssl/issues/1072, that it was fixed, or it re-appeared because of some reason. Since `/Q` switch is not enough to silence errors, the possible solution is to replace `DEL ...dll` commands with condition `if exist ...dll (DEL ...dll)`.

Hmmm, can't recall that I promised a fix then, but I can promise it now ;-)  Will fix.

> Some of issues above could describe expected behavior. In such case no need to reply on them.

**Comment by sav-ix** [comment]
>>Mingw-w64 installation contain files:
>>...
>>which are missing in MSVC installation,

> Yup. Most of those are shell scripts, so completely unusable on native Windows. CA.pl is the exception, but is there a way to associate it with perl so it becomes easy to use on the command line?

ActivePerl associate itself with `.pl` files during installation, and `CA.pl` could be run from Windows Command Prompt even without PERL in PATH variable. Though need to be verified, whether it works correctly.

>>Some files in html docs of Mingw-w64 installation contain plain text, eg:
>>...
>>while the same text in html docs of MSVC installation would be:

> That's a difference that we don't control. I believe it comes down to different perl versions, or that some perl dists add the extra <span> tags 

Or because ActivePerl installation contain modified `pod2html` package.

>> Noise in stderr still exists:
...

> Hmmm, can't recall that I promised a fix then, but I can promise it now ;-) Will fix.

It was a part of PR: https://github.com/openssl/openssl/commit/592b6fb489c97be2b039ab671647aa58a472204a ; see also https://github.com/openssl/openssl/issues/1072#issuecomment-219369961 .

**Comment by levitte** [comment]
>>>Mingw-w64 installation contain files:
>>>...
>>>which are missing in MSVC installation,

>> Yup. Most of those are shell scripts, so completely unusable on native Windows. CA.pl is the exception, but is there a way to associate it with perl so it becomes easy to use on the command line?

>ActivePerl associate itself with `.pl` files during installation, and `CA.pl` could be run from Windows Command Prompt even without PERL in PATH variable. Though need to be verified, whether it works correctly.

Ok cool.  I gotta try that out :-)

>>>Some files in html docs of Mingw-w64 installation contain plain text, eg:
>>>...
>>>while the same text in html docs of MSVC installation would be:

>> That's a difference that we don't control. I believe it comes down to different perl versions, or that some perl dists add the extra <span> tags 

>Or because ActivePerl installation contain modified `pod2html` package.

That's what I meant by refering to "some perl dists"...

>>> Noise in stderr still exists:
>...

>> Hmmm, can't recall that I promised a fix then, but I can promise it now ;-) Will fix.

>It was a part of PR: https://github.com/openssl/openssl/commit/592b6fb489c97be2b039ab671647aa58a472204a ; see also https://github.com/openssl/openssl/issues/1072#issuecomment-219369961 .

Okie.  Either way, will fix.

**Comment by levitte** [comment]
So in summary, here are the issues you presented:

-    some shell scripts not installed on native Windows.  *Will not fix, except for CA.pl*
-    HTML docs differing in unix/mingw installs and on native Windows.  *Fixed*  (with thanks, this uncovered some faults in the OCSP .pod files)
-    Title missing.  *Fixed*
-    Different code block HTML.  *Will not fix, this depends on perl dists*
-    Native Windows builds won't use parallellism to build the docs.  *Will not fix at this stage, but for a later OpenSSL version*
-    Shut DEL up regarding non-existent files.  *Will fix*

**Comment by levitte** [comment]
#1107 should clear up the documentation issues, both on unix/mingw and on native Windows.

**Comment by sav-ix** [comment]
Thought to test https://github.com/openssl/openssl/pull/1107 , but currently OpenSSL master seems to be broken for both Mingw-w64:
```
./Configure --prefix="$PWD/../GCC64d" -std=gnu11 -mtune=native -march=native threads no-deprecated shared debug-mingw64 >_Configure.log && ma
ke

[snip]

gcc -DDSO_WIN32 -DOPENSSL_THREADS -DOPENSSL_NO_STATIC_ENGINE -DOPENSSL_PIC -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DPOLY1305_ASM -DOPENSSL_API_COMPAT=0x10100000L -DOPENSSLDIR="\"/c:/libOPENSSL-1.1.0-dev/src/../GCC64d/ssl\"" -DENGINESDIR="\"/c:/libOPENSSL-1.1.0-dev/src/../GCC64d/lib/engines\"" -DL_ENDIAN -DWIN32_LEAN_AND_MEAN -DUNICODE -D_UNICODE -m64 -Wall -g -O0 -D_MT  -std=gnu11 -mtune=native -march=native -D_WINDLL -Iinclude -I. -Icrypto/include -MMD -MF crypto/lhash/lh_stats.d.tmp -MT crypto/lhash/lh_stats.o -c -o crypto/lhash/lh_stats.o crypto/lhash/lh_stats.c
In file included from crypto/lhash/lh_stats.c:21:0:
crypto/lhash/lhash_lcl.h:18:5: error: unknown type name 'LHASH_NODE'
     LHASH_NODE **b;
     ^
crypto/lhash/lh_stats.c: In function 'OPENSSL_LH_node_stats_bio':
crypto/lhash/lh_stats.c:89:16: warning: assignment from incompatible pointer type [-Wincompatible-pointer-types]
         for (n = lh->b[i], num = 0; n != NULL; n = n->next)
                ^
crypto/lhash/lh_stats.c: In function 'OPENSSL_LH_node_usage_stats_bio':
crypto/lhash/lh_stats.c:103:16: warning: assignment from incompatible pointer type [-Wincompatible-pointer-types]
         for (n = lh->b[i], num = 0; n != NULL; n = n->next)
                ^
Makefile:3860: recipe for target 'crypto/lhash/lh_stats.o' failed
make: *** [crypto/lhash/lh_stats.o] Error 1

```

and MSVC:
```
perl Configure --prefix="%CD%\..\MSVC64r\dll" --openssldir="%CD%\..\MSVC64r\dll\ssl" -EHsc -arch:AVX -FS threads no-deprecated shared VC-WIN64A && nmake

[snip]

        cl -DOPENSSL_USE_APPLINK -DDSO_WIN32 -DNDEBUG -DOPENSSL_THREADS -DOPENSSL_NO_STATIC_ENGINE -DOPENSSL_PIC -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DPOLY1305_ASM -DOPENSSL_API_COMPAT=0x10100000L "-DENGINESDIR=\"c:\\libOPENSSL-1.1.0-dev\\MSVC64r\\dll\\lib\\engines\"" "-DOPENSSLDIR=\"c:\\libOPENSSL-1.1.0-dev\\src\\..\\MSVC64r\\dll\\ssl\"" -W3 -wd4090 -Gs0 -GF -Gy -nologo -DOPENSSL_SYS_WIN32 -DWIN32_LEAN_AND_MEAN -DL_ENDIAN -D_CRT_SECURE_NO_DEPRECATE -DUNICODE -D_UNICODE /MD /O2  -EHsc -arch:AVX -FS /Zi /Fdossl_static /I include /I . /I crypto\include -c /Focrypto\lhash\lh_stats.obj crypto\lhash\lh_stats.c
lh_stats.c
c:\libopenssl-1.1.0-dev\src\crypto\lhash\lhash_lcl.h(18): error C2061: syntax error: identifier 'LHASH_NODE'
c:\libopenssl-1.1.0-dev\src\crypto\lhash\lhash_lcl.h(42): error C2059: syntax error: '}'
crypto\lhash\lh_stats.c(64): error C2037: left of 'num_items' specifies undefined struct/union 'lhash_st'
crypto\lhash\lh_stats.c(65): error C2037: left of 'num_nodes' specifies undefined struct/union 'lhash_st'
crypto\lhash\lh_stats.c(66): error C2037: left of 'num_alloc_nodes' specifies undefined struct/union 'lhash_st'
crypto\lhash\lh_stats.c(67): error C2037: left of 'num_expands' specifies undefined struct/union 'lhash_st'
crypto\lhash\lh_stats.c(68): error C2037: left of 'num_expand_reallocs' specifies undefined struct/union 'lhash_st'
crypto\lhash\lh_stats.c(69): error C2037: left of 'num_contracts' specifies undefined struct/union 'lhash_st'
crypto\lhash\lh_stats.c(71): error C2037: left of 'num_contract_reallocs' specifies undefined struct/union 'lhash_st'
crypto\lhash\lh_stats.c(72): error C2037: left of 'num_hash_calls' specifies undefined struct/union 'lhash_st'
crypto\lhash\lh_stats.c(73): error C2037: left of 'num_comp_calls' specifies undefined struct/union 'lhash_st'
crypto\lhash\lh_stats.c(74): error C2037: left of 'num_insert' specifies undefined struct/union 'lhash_st'
crypto\lhash\lh_stats.c(75): error C2037: left of 'num_replace' specifies undefined struct/union 'lhash_st'
crypto\lhash\lh_stats.c(76): error C2037: left of 'num_delete' specifies undefined struct/union 'lhash_st'
crypto\lhash\lh_stats.c(77): error C2037: left of 'num_no_delete' specifies undefined struct/union 'lhash_st'
crypto\lhash\lh_stats.c(78): error C2037: left of 'num_retrieve' specifies undefined struct/union 'lhash_st'
crypto\lhash\lh_stats.c(79): error C2037: left of 'num_retrieve_miss' specifies undefined struct/union 'lhash_st'
crypto\lhash\lh_stats.c(80): error C2037: left of 'num_hash_comps' specifies undefined struct/union 'lhash_st'
crypto\lhash\lh_stats.c(88): error C2037: left of 'num_nodes' specifies undefined struct/union 'lhash_st'
crypto\lhash\lh_stats.c(89): error C2037: left of 'b' specifies undefined struct/union 'lhash_st'
crypto\lhash\lh_stats.c(102): error C2037: left of 'num_nodes' specifies undefined struct/union 'lhash_st'
crypto\lhash\lh_stats.c(103): error C2037: left of 'b' specifies undefined struct/union 'lhash_st'
crypto\lhash\lh_stats.c(110): error C2037: left of 'num_nodes' specifies undefined struct/union 'lhash_st'
crypto\lhash\lh_stats.c(115): error C2037: left of 'num_nodes' specifies undefined struct/union 'lhash_st'
crypto\lhash\lh_stats.c(116): error C2037: left of 'num_nodes' specifies undefined struct/union 'lhash_st'
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\amd64\cl.EXE"' : return code '0x2'
Stop.
```

**Comment by levitte** [comment]
Here's the quick fix for that (to be committed):

``` diff
diff --git a/crypto/lhash/lhash_lcl.h b/crypto/lhash/lhash_lcl.h
index cfb0e1a..eb4a1a3 100644
--- a/crypto/lhash/lhash_lcl.h
+++ b/crypto/lhash/lhash_lcl.h
@@ -15,7 +15,7 @@ struct lhash_node_st {
 };
 
 struct lhash_st {
-    LHASH_NODE **b;
+    OPENSSL_LH_NODE **b;
     OPENSSL_LH_COMPFUNC comp;
     OPENSSL_LH_HASHFUNC hash;
     unsigned int num_nodes;
```

**Comment by sav-ix** [comment]
Tested your OpenSSL fork https://github.com/levitte/openssl/tree/misc-fix-docs-20160521-1
with `lhash` fix applied:

- `lhash` build errors gone,
- the only difference with html docs, that file
```
share\doc\openssl\html\man1\c_rehash.html
```

exist only in Mingw-w64 installation, and files
```
html\man1\contributedundertheOpenSSLlicense.c_rehash.html
html\man1\OriginaltextbyJamesWestby.html
```
exist only in MSVC installation.

**Comment by levitte** [comment]
Try again.

**Comment by levitte** [comment]
Or, well, it just got pushed to master

**Comment by sav-ix** [comment]
Tested recent master. The "html docs" issue seems to be closed.

**Comment by levitte** [comment]
Cool, so currently, the state is this:

-    [x] some shell scripts not installed on native Windows.  *Will not fix, except for CA.pl*
-    [x] HTML docs differing in unix/mingw installs and on native Windows.  *Fixed*  (with thanks, this uncovered some faults in the OCSP .pod files)
-    [x] Title missing.  *Fixed*
-    [x] Different code block HTML.  *Will not fix, this depends on perl dists*
-    [x] Native Windows builds won't use parallellism to build the docs.  *Will not fix at this stage, but for a later OpenSSL version*
-    [ ] Shut DEL up regarding non-existent files.  *Will fix*

**Comment by sav-ix** [comment]
- I don't see `CA.pl` in MSVC installation,
- as for parallelism for MSVC docs build, going to open separate issue with short feature request list for OpenSSL-1.2.0.

**Comment by levitte** [comment]
You're right, I forgot about CA.pl for a moment.  Unchecked that item

**Comment by levitte** [comment]
As for a feature request list, I suggest one issue per feature, and label them with the appropriate milestone (1.2.0).  That makes them easier to handle, and especially to spread the load.

**Comment by levitte** [comment]
Shutting up DEL done in #1109

**Comment by sav-ix** [comment]
PR: https://github.com/openssl/openssl/pull/1109 works fine. no more `DEL` noise.

**Comment by levitte** [comment]
It's in master now.

**Comment by levitte** [comment]
Ok, so those scripts are in focus now, see #1112

**Comment by sav-ix** [comment]
Tested PR: https://github.com/openssl/openssl/pull/1112 from [your fork](https://github.com/levitte/openssl/tree/fix-installed-scripts-20160523-1).
Files:
```
bin\c_rehash.pl
ssl\misc\CA.pl
ssl\misc\tsget.pl
```
appeared in OpenSSL MSVC installation.

**Comment by sav-ix** [comment]
Got some additions to 1st post (could be reported as separate issue, if needed):
- [ ] For builds using Mingw-w64, if re-run `mingw32-make` after all binaries were built, got error:
```
sh-4.3$ mingw32-make
/usr/bin/sh: -c: line 0: unexpected EOF while looking for matching ``'
/usr/bin/sh: -c: line 1: syntax error: unexpected end of file
Makefile:172: recipe for target 'depend' failed
mingw32-make: *** [depend] Error 1
```
That's a bit differs from autotools build systems, which, in such case, process `Makefile`, find, that all binaries already built, and report, that all jobs done.

And no such errors for builds using MSVC after `nmake` re-run.


- [ ] Inconsistencies for non `--prefix=` builds:

- for builds using Mingw-w64, installation of OpenSSL binaries and data made to "virtual" filesystem, which MSYS2 trying to emulate. That would be `<MSYS2_installation>\usr\local` directory:
```
sh-4.3$ mingw32-make install

[snip]

created directory `/usr/local'
created directory `/usr/local/lib'
created directory `/usr/local/lib/engines'
created directory `/usr/local/bin'
created directory `/usr/local/include'
created directory `/usr/local/include/openssl'
*** Installing engines
created directory `/usr/local/ssl'
created directory `/usr/local/ssl/misc'
```
- for builds using MSVC, installation of OpenSSL binaries made to `%ProgramFiles%\OpenSSL` or `$ProgramFiles(x86)%\OpenSSL` and data to `%CommonProgramFiles%\SSL` folders.


- [ ] Installation of non `--prefix=` builds using MSVC on Windows 10 finishes with errors:
```
c:\libOPENSSL-1.1.0-dev\src\nmake install

[snip]

Microsoft (R) Program Maintenance Utility Version 14.00.23506.0
Copyright (C) Microsoft Corporation.  All rights reserved.

*** Installing development files
Cannot create directory C:/Program Files/OpenSSL: Permission denied
NMAKE : fatal error U1077: 'C:\ProgramData\Perl64\bin\perl.exe' : return code '0xd'
Stop.
```
if there is no Administrative privileges for current Windows Command Prompt instance. Could make sense to add such info to `INSTALL` manual.


- [ ] Inconsistencies for `--prefix=` builds:
- for builds using Mingw-w64 no need to set `--openssldir=` key, since OpenSSL data area automatically set to `--prefix=...\ssl`,
- for builds using MSVC, if `--openssldir=` key wasn't set, OpenSSL data installed into `%CommonProgramFiles%\SSL` folder. Is last rule a suspected behavior for `--prefix=` builds?

**Comment by levitte** [comment]
#1112 pushed to master

**Comment by levitte** [comment]
> Got some additions to 1st post (could be reported as separate issue, if needed):

> - [ ] For builds using Mingw-w64, if re-run `mingw32-make` after all binaries were built, got error:
> ```
sh-4.3$ mingw32-make
/usr/bin/sh: -c: line 0: unexpected EOF while looking for matching ``'
/usr/bin/sh: -c: line 1: syntax error: unexpected end of file
Makefile:172: recipe for target 'depend' failed
mingw32-make: *** [depend] Error 1
```

> That's a bit differs from autotools build systems, which, in such case, process `Makefile`, find, that all binaries already built, and report, that all jobs done.

Hmmm...

> And no such errors for builds using MSVC after `nmake` re-run.

Correct, `depend` is currently a no-op there...

> - [ ] Inconsistencies for non `--prefix=` builds:

> - for builds using Mingw-w64, installation of OpenSSL binaries and data made to "virtual" filesystem, which MSYS2 trying to emulate. That would be `<MSYS2_installation>\usr\local` directory:

> ```
sh-4.3$ mingw32-make install

> [snip]

> created directory `/usr/local'
created directory `/usr/local/lib'
created directory `/usr/local/lib/engines'
created directory `/usr/local/bin'
created directory `/usr/local/include'
created directory `/usr/local/include/openssl'
*** Installing engines
created directory `/usr/local/ssl'
created directory `/usr/local/ssl/misc'
```

> - for builds using MSVC, installation of OpenSSL binaries made to `%ProgramFiles%\OpenSSL` or `$ProgramFiles(x86)%\OpenSSL` and data to `%CommonProgramFiles%\SSL` folders.

Yes.  Or well, "data" is viewed as a separate structure that can survive through many OpenSSL versions.  A unix-like environment is fundamentally different from a native Windows one, so we adapt to the platform we build for.

> - [ ] Installation of non `--prefix=` builds using MSVC on Windows 10 finishes with errors:

> ```
c:\libOPENSSL-1.1.0-dev\src\nmake install

> [snip]

> Microsoft (R) Program Maintenance Utility Version 14.00.23506.0
Copyright (C) Microsoft Corporation.  All rights reserved.

> *** Installing development files
Cannot create directory C:/Program Files/OpenSSL: Permission denied
NMAKE : fatal error U1077: 'C:\ProgramData\Perl64\bin\perl.exe' : return code '0xd'
Stop.
```

> if there is no Administrative privileges for current Windows Command Prompt instance. Could make sense to add such info to `INSTALL` manual.

Yeah, I only just discovered this myself and thought that it was something wrong with my setup.  I found it weird, since my user *is* in the Administrators group...  if you have a solution for enabling privs, I'm all ears.

> - [ ] Inconsistencies for `--prefix=` builds:

> - for builds using Mingw-w64 no need to set `--openssldir=` key, since OpenSSL data area automatically set to `--prefix=...\ssl`,
> - for builds using MSVC, if `--openssldir=` key wasn't set, OpenSSL data installed into `%CommonProgramFiles%\SSL` folder. Is last rule a suspected behavior for `--prefix=` builds?

Yeah, the Unix environment is a bit magic in that sense.  But then again, Unix does have the "super-directory" `/usr` or `/usr/local` while Windows does have the separate `%CommonProgramFiles%` for the purpose of data files.  Like I said, we adapt to the platform we build for.

I did consider *not* having the magic building of openssldir on top of prefix on Unix, but decided to keep it for the time being, for backward compatibility.  For Windows and VMS, there's no comparable precedent.

-----

So in summary, there is the priviliges problem in the command prompt to deal with.

**Comment by levitte** [comment]
Oh, and that `depend` issue...

**Comment by levitte** [comment]
Too see what's happening with `depend`, apply this patch, configure and make and again so you get that same error, and show me what you got.

``` diff
diff --git a/Configurations/unix-Makefile.tmpl b/Configurations/unix-Makefile.tmpl
index 469bd32..fb4c3f2 100644
--- a/Configurations/unix-Makefile.tmpl
+++ b/Configurations/unix-Makefile.tmpl
@@ -269,7 +269,7 @@ clean: libclean
 # concatenate only if that is true.
 depend:
 	@: {- output_off() if $disabled{makedepend}; "" -}
-	@if [ -n "`find $(DEPS) -newer Makefile 2>/dev/null; exit 0`" ]; then \
+	if [ -n "`find $(DEPS) -newer Makefile 2>/dev/null; exit 0`" ]; then \
 	  ( sed -e '/^# DO NOT DELETE THIS LINE.*/,$$d' < Makefile; \
 	    echo '# DO NOT DELETE THIS LINE -- make depend depends on it.'; \
 	    echo; \
```

**Comment by levitte** [comment]
I've addressed the privs issue in #1114

**Comment by levitte** [comment]
(#1114 has just been merged into master)

**Comment by sav-ix** [comment]
> if you have a solution for enabling privs, I'm all ears.

PR: https://github.com/openssl/openssl/pull/1114/commits/29bbf703259e198531ac3142e94717105b70dd21 seems to be the best solution. Using `runas` or 3rd party stuff, like `sudo`, `elevate`, etc., has its own drawbacks and better be avoided.

>  apply this patch, configure and make and again so you get that same error

Got more verbose output:
```

sh-4.3$ mingw32-make
if [ -n "`find test/sha512t.d 

[snip]

 crypto/evp/c_allc.d crypto/evp/m_md5.d -newer Makefile 2>/dev/null; exit 0`" ]; then \
  ( sed -e '/^# DO NOT DELETE THIS LINE.*/,$d' < Makefile; \
    echo '# DO NOT DELETE THIS LINE -- make depend depends on it.'; \
    echo; \
    for f in test/sha512t.d 

[snip]
	
	crypto/evp/m_md5.d; do \
      if [ -f $f ]; then cat $f; fi; \
    done ) > Makefile.new; \
  if cmp Makefile.new Makefile >/dev/null 2>&1; then \
    rm -f Makefile.new; \
  else \
    mv -f Makefile.new Makefile; \
  fi; \
fi
/usr/bin/sh: -c: line 0: unexpected EOF while looking for matching ``'
/usr/bin/sh: -c: line 1: syntax error: unexpected end of file
Makefile:172: recipe for target 'depend' failed
mingw32-make: *** [depend] Error 1
```

Just in case, error occur during re-run of `mingw32-make` command, **not** `mingw32-make depend`.
Though `mingw32-make depend` throw same error.

**Comment by levitte** [comment]
Ok, the only thing I can understand from that is that the ```if [ -n "`find ...``` line got chopped into two lines for some reason.  Why, I cannot fathom.  I just tried in my Mings32 shell, and it worked fine (it's a standard MSYS2/Mingw installation, straight from the web page).

So, uhmmm, cannot reproduce.

**Comment by sav-ix** [comment]
Added [full log.zip](https://github.com/openssl/openssl/files/279481/mingw-w64_make_depend_error_log.zip) in attachment, in case that help.

BTW, for some other libraries, build errors occur, when parameter list for Mingw-w64 linker exceed 10.000 symbols limit (Windows feature; AFAIR, 32.768 for Linux). Not sure, whether this is the same case, but during `make depend` run `find` command receive parameters list over 15.000 symbols long.

But even so, it does not explain, why this error not reproduce on your system.

**Comment by levitte** [comment]
With "symbols", do you mean "characters on the command line"?  Like, something like `apps/openssl.c` is 14 "symbols"?

**Comment by sav-ix** [comment]
yes.

**Comment by levitte** [comment]
Ok...  Since it's the shell that complains, it's also the shell that somehow limited what there was to limit.  From the error messages, it sounds like it simply decided to truncate the line.  What shell are you using?  I'm using bash...

**Comment by sav-ix** [comment]
Using MSYS2 x64. Run `mingw64_shell.bat` to launch shell.

And already found source of error: run `mingw32-make` (from Mingw-w64) instead of `make` (from MSYS2). My bad. Sorry for the noise.

Thus all questions in this issue are closed.

**Comment by levitte** [comment]
All right then.  Thank you!

**Comment by sav-ix** [comment]
Not at all. Than you for fixes.
